R/msgmerge.R

```diff
@@ -31,28 +35,69 @@ run_msgfmt = function(po_file, mo_file, verbose) {
   return(invisible())
 }
 
-update_mo_files = function(dir, package, verbose) {
-  inst_dir <- file.path(dir, "inst", "po")
-  lang_regex <- "^(R-)?([a-zA-Z_]+)\\.po$"
+#' Compile `.po` files to `.mo`
+#'
+#' This function compiles the plain text `.po` files that translators work with
+#' into the binary `.mo` files that are installed with packages and used for
+#' live translations.
+#'
+#' @param dir Path to package root directory.
+#' @param package Name of package. If not supplied, read from `DESCRIPTION`.
+#' @param lazy If `TRUE`, only `.mo` functions that are older than `.po`
+#'   files be updated
+#' @param verbose If `TRUE`, print information as it goes.
+po_compile = function(dir = ".", package = NULL, lazy = TRUE, verbose = TRUE) {
+  po_files <- po_files(dir = dir, package = package, lazy = lazy)
+  dir_create(dirname(po_files$mo))
 
-  po_files <- list.files(file.path(dir, "po"), pattern = "\\.po$")
-  languages <- gsub(lang_regex, "\\2", po_files)
-  mo_files <- gsub(lang_regex, sprintf("\\1%s.mo", package), po_files)
-  mo_dirs <- file.path(inst_dir, languages, "LC_MESSAGES")
-  # NB: dir.create() only accepts one directory at a time...
-  for (mo_dir in unique(mo_dirs)) dir.create(mo_dir, recursive = TRUE, showWarnings = FALSE)
+  if (!lazy) {
+    # Clear out all older translations
+    mo_dirs <- dir(file.path(dir, "inst", "po"), full.names = TRUE)
+    to_delete <- mo_dirs[!basename(mo_dirs) %in% c(po_files$language, "en@quot")]
+
+    for (dir in to_delete) {
+      if (verbose) messagef("Deleting unmatched %s translation", basename(dir))
+      unlink(dir, recursive = TRUE)
+    }
+  }
 
-  for (ii in seq_along(po_files)) {
+  for (ii in seq_len(nrow(po_files))) {
+    if (verbose) messagef("Recompiling %s translation", po_files$language[ii])
     run_msgfmt(
-      po_file = file.path(dir, "po", po_files[ii]),
-      mo_file = file.path(mo_dirs[ii], mo_files[ii]),
+      po_file = po_files$po[ii],
+      mo_file = po_files$mo[ii],
       verbose = verbose
     )
   }
 
   return(invisible())
 }
 
+po_files <- function(dir = ".", package = NULL, lazy = TRUE) {
+  if (is.null(package)) {
+    package <- get_desc_data(dir)[["Package"]]
+  }
+
+  po_paths <- list.files(file.path(dir, "po"), pattern = "\\.po$", full.names = TRUE)
+
+  lang_regex <- "^(R-)?([a-z]{2}(?:_[A-Z]{2})?)\\.po$"
+  languages <- gsub(lang_regex, "\\2", basename(po_paths))
+  mo_names <- gsub(lang_regex, sprintf("\\1%s.mo", package), basename(po_paths))
+  mo_paths <- file.path(dir, "inst", "po", languages, "LC_MESSAGES", mo_names)
+
+  out <- data.table(
+    language = languages,
+    po = po_paths,
+    mo = mo_paths
+  )
+
+  if (lazy) {
```

Comment:
i'm thinking it's cleaner to leave `lazy` out of the logic here and do this part in `po_compile` directly, WDYT?

Reply:
Yeah, makes sense to me.